src/Workspaces/Core/Portable/Options/DocumentOptionSet.cs

```diff
@@ -35,9 +36,12 @@ public override OptionSet WithChangedOption(OptionKey optionAndLanguage, object
             return new DocumentOptionSet(_backingOptionSet.WithChangedOption(optionAndLanguage, value), _language);
         }
 
-        internal override IEnumerable<OptionKey> GetChangedOptions(OptionSet optionSet)
+        IEnumerable<OptionKey> IInternalOptionSet.GetChangedOptions(OptionSet optionSet)
         {
-            return _backingOptionSet.GetChangedOptions(optionSet);
+            if (!(_backingOptionSet is IInternalOptionSet otherOptionSet))
```

Comment:
Why not enforce this by making the ctor parameter take an `IInternalOptionSet` parameter?